sdks/java/extensions/euphoria/euphoria-beam/src/main/java/org/apache/beam/sdk/extensions/euphoria/beam/BrodcastHashJoinTranslator.java

```diff
@@ -0,0 +1,177 @@
+public class BrodcastHashJoinTranslator implements OperatorTranslator<Join> {
+
+  public static boolean hasFitsInMemoryHint(Operator operator) {
+    return operator != null
+        && operator.getHints() != null
+        && operator.getHints().contains(SizeHint.FITS_IN_MEMORY);
+  }
+
+  @Override
+  public PCollection<?> translate(Join operator, BeamExecutorContext context) {
+    return doTranslate(operator, context);
+  }
+
+  public <K, LeftT, RightT, OutputT, W extends Window<W>> PCollection<Pair<K, OutputT>> doTranslate(
+      Join<LeftT, RightT, K, OutputT, W> operator, BeamExecutorContext context) {
+    Coder<K> keyCoder = context.getCoder(operator.getLeftKeyExtractor());
+
+    List<PCollection<Object>> inputs = context.getInputs(operator);
+    final PCollection<Object> left = inputs.get(0);
+    final PCollection<Object> right = inputs.get(1);
+    final PCollection<KV<K, LeftT>> leftKvInput =
+        getKVInputCollection(
+            left,
+            operator.getLeftKeyExtractor(),
+            keyCoder,
+            new KryoCoder<>(),
+            ":extract-keys-left");
+
+    final PCollection<KV<K, RightT>> rightKvInput =
+        getKVInputCollection(
+            right,
+            operator.getRightKeyExtractor(),
+            keyCoder,
+            new KryoCoder<>(),
+            ":extract-keys-right");
+
+    switch (operator.getType()) {
+      case LEFT:
+        final PCollectionView<Map<K, Iterable<RightT>>> broadcastRight =
+            rightKvInput.apply(View.asMultimap());
+        return leftKvInput.apply(
+            ParDo.of(new BroadcastHashLeftJoinFn<>(broadcastRight, operator.getJoiner()))
+                .withSideInputs(broadcastRight));
+
+      case RIGHT:
+        final PCollectionView<Map<K, Iterable<LeftT>>> broadcastLeft =
+            leftKvInput.apply(View.asMultimap());
+        return rightKvInput.apply(
+            ParDo.of(new BroadcastHashRightJoinFn<>(broadcastLeft, operator.getJoiner()))
+                .withSideInputs(broadcastLeft));
+
+      default:
+        throw new UnsupportedOperationException(
+            String.format(
+                "Cannot translate Euphoria '%s' operator to Beam transformations."
+                    + " Given join type '%s' is not supported for BrodcastHashJoin.",
+                Join.class.getSimpleName(), operator.getType()));
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean canTranslate(Join operator) {
+    final ArrayList<Dataset> inputs = new ArrayList(operator.listInputs());
+    if (inputs.size() != 2) {
+      return false;
+    }
+    final Dataset leftDataset = inputs.get(0);
+    final Dataset rightDataset = inputs.get(1);
+    return (operator.getType() == Join.Type.LEFT && hasFitsInMemoryHint(rightDataset.getProducer())
+            || operator.getType() == Join.Type.RIGHT
+                && hasFitsInMemoryHint(leftDataset.getProducer()))
+        && !(operator.getWindowing() instanceof MergingWindowing);
```

Comment:
You cannot expect Euphoria windowing here, since it cannot be translated into Beam windowing. We use `BeamWindowing` envelope to define Beam windowing through Euphoria API.

Reply:
👍